larpandoracontent/LArTrackShowerId/TrackShowerIdFeatureTool.cc

```diff
@@ -298,6 +298,140 @@ StatusCode PfoHierarchyFeatureTool::ReadSettings(const TiXmlHandle /*xmlHandle*/
+
+        concentration = (chargeCore + chargeHalo > std::numeric_limits<float>::epsilon()) ? chargeCon / (chargeCore + chargeHalo) : -1.0;
+        
+        conicalness = this->CalculateConicalness(clusterCaloHitList, pfoStart, eigenVecs[0], pfoLength);
+    }
+
+    featureVector.push_back(haloTotalRatio);
+    featureVector.push_back(concentration);
+    featureVector.push_back(conicalness);
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+void ConeShapeFeatureTool::CalculateChargeDistribution(const CaloHitList caloHitList, const CartesianVector &pfoStart, const CartesianVector &pfoDir, 
+	float &chargeCore, float &chargeHalo, float &chargeCon)
+{
+    for (const CaloHit *const pCaloHit : caloHitList)
+    {
+        float distFromTrackFit = ((pCaloHit->GetPositionVector() - pfoStart).GetCrossProduct(pfoDir)).GetMagnitude();
+
+        if (distFromTrackFit < m_MoliereRadiusFrac * m_MoliereRadius)
+            chargeCore += pCaloHit->GetInputEnergy();
+        else
+            chargeHalo += pCaloHit->GetInputEnergy();
+        
+        chargeCon += pCaloHit->GetInputEnergy() / std::max(1.E-2f, distFromTrackFit);
+    }
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+float ConeShapeFeatureTool::CalculateConicalness(const CaloHitList &caloHitList, const CartesianVector &pfoStart, const CartesianVector &pfoDir, const float pfoLength)
+{
+
+    float totalChargeStart(0.f), totalChargeEnd(0.f);
+    float chargeConStart(0.f), chargeConEnd(0.f);
+    unsigned int nHitsConStart(0), nHitsConEnd(0);
+    
+    for (const CaloHit *const pCaloHit : caloHitList)
+    {
+        float distFromTrackFit = ((pCaloHit->GetPositionVector() - pfoStart).GetCrossProduct(pfoDir)).GetMagnitude();
+        float hitLength = std::fabs((pCaloHit->GetPositionVector() - pfoStart).GetDotProduct(pfoDir)); 
+ 
+        if (hitLength / pfoLength < m_conFracRange) 
+        {
+            chargeConStart += distFromTrackFit * distFromTrackFit * pCaloHit->GetInputEnergy();
+            ++nHitsConStart;
+            totalChargeStart += pCaloHit->GetInputEnergy();
+        }
+        else if (1. - hitLength / pfoLength < m_conFracRange) 
+        {
+            chargeConEnd += distFromTrackFit * distFromTrackFit * pCaloHit->GetInputEnergy();
+            ++nHitsConEnd;
+            totalChargeEnd += pCaloHit->GetInputEnergy();
+        }
+    }
+
+    float conicalness(-1.f);
+    
+    if (nHitsConStart >= m_conMinHits && nHitsConEnd >= m_conMinHits && totalChargeEnd > m_minCharge && 
+   	std::sqrt(chargeConStart) > m_minCharge && totalChargeStart > m_minCharge) 
+        conicalness = (std::sqrt(chargeConEnd) / totalChargeEnd) / (std::sqrt(chargeConStart) / totalChargeStart);
```

Comment:
The efficiency could be improved here by only using a single square root

sqrt(chargeConEnd / chargeConStart) * (totalChargeStart / totalChargeEnd)

Reply:
There may need to be additional checks here to ensure no division by 0?